// @sanjo/mongodb-diff/dist/lib/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});

var _stringify = require('babel-runtime/core-js/json/stringify');

var _stringify2 = _interopRequireDefault(_stringify);

var _keys = require('babel-runtime/core-js/object/keys');

var _keys2 = _interopRequireDefault(_keys);

exports.diff = diff;

var _deepDiff = require('deep-diff');

var _lodash = require('lodash');

var _lodash2 = _interopRequireDefault(_lodash);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

function getProperty(object, path) {
  return _lodash2.default.get(object, path);
}

function getPropertyPathAsString(changeDiff) {
  return pathToString(changeDiff.path);
}

function pathToString(path) {
  return path.join('.');
}

function doesPreferredSetOperationExist(update, path) {
  if (update.$set) {
    for (var i = 1; i <= path.length; i += 1) {
      if (update.$set[pathToString(path.slice(0, i))]) {
        return true;
      }
    }
  }

  return false;
}

function isArrayIndexPath(path) {
  return _lodash2.default.isNumber(_lodash2.default.last(path));
}

function isPropertyRemove(change) {
  return change.diff.kind === 'D';
}

function isPropertyUpdate(change) {
  return (change.diff.kind === 'E' || change.diff.kind === 'N') && !isArrayIndexPath(change.diff.path);
}

function isArrayPush(change) {
  return change.diff.kind === 'A' && change.diff.item.kind === 'N';
}

function isArrayPull(change) {
  if (change.diff.kind === 'A' && change.diff.item.kind === 'D') {
    var array = getProperty(change.rhs, change.diff.path);
    if (!array || !_lodash2.default.includes(array, change.diff.item.lhs)) {
      return true;
    }
  }

  return false;
}

function isArrayUpdate(change) {
  return change.diff.kind === 'A' || (change.diff.kind === 'E' || change.diff.kind === 'N') && isArrayIndexPath(change.diff.path);
}

function setProperty(update, propertyPath, value) {
  if (!update.$set) {
    update.$set = {};
  }
  update.$set[propertyPath] = value;
}

function unsetProperty(update, propertyPath) {
  if (!update.$unset) {
    update.$unset = {};
  }
  update.$unset[propertyPath] = true;
}

function push(update, propertyPath, value) {
  if (!update.$push) {
    update.$push = {};
  }
  if (!update.$push[propertyPath]) {
    update.$push[propertyPath] = value;
  } else {
    if (update.$push[propertyPath].$each) {
      update.$push[propertyPath].$each.push(value);
    } else {
      update.$push[propertyPath] = {
        $each: [update.$push[propertyPath], value]
      };
    }
  }
}

function pull(update, propertyPath, value) {
  if (!doesPullPropertyExist(update, propertyPath) && !doesPullAllPropertyExist(update, propertyPath)) {
    if (!update.$pull) {
      update.$pull = {};
    }
    update.$pull[propertyPath] = value;
  } else {
    if (!update.$pullAll) {
      update.$pullAll = {};
    }
    if (update.$pullAll[propertyPath]) {
      update.$pullAll[propertyPath].push(value);
    } else {
      update.$pullAll[propertyPath] = [update.$pull[propertyPath], value];
      delete update.$pull[propertyPath];
      if ((0, _keys2.default)(update.$pull).length === 0) {
        delete update.$pull;
      }
    }
  }
}

function doesPullPropertyExist(update, propertyPath) {
  return update.$pull && update.$pull[propertyPath];
}

function doesPullAllPropertyExist(update, propertyPath) {
  return update.$pullAll && update.$pullAll[propertyPath];
}

function createChangeHandler(check, updater) {
  return function changeHandler(update, change) {
    if (check(change)) {
      updater(update, change);
      return true;
    }
  };
}

function applyChangeHandlers(changeHandlers, update, change) {
  _lodash2.default.every(changeHandlers, function (changeHandler) {
    return !changeHandler(update, change);
  });
}

var changeHandlers = [createChangeHandler(isPropertyUpdate, function (update, change) {
  setProperty(update, getPropertyPathAsString(change.diff), change.diff.rhs);
}), createChangeHandler(isPropertyRemove, function (update, change) {
  unsetProperty(update, getPropertyPathAsString(change.diff));
}), createChangeHandler(isArrayPush, function (update, change) {
  if (!doesPreferredSetOperationExist(update, change.diff.path)) {
    push(update, getPropertyPathAsString(change.diff), change.diff.item.rhs);
  }
}), createChangeHandler(isArrayPull, function (update, change) {
  if (!doesPreferredSetOperationExist(update, change.diff.path)) {
    pull(update, getPropertyPathAsString(change.diff), change.diff.item.lhs);
  }
}), createChangeHandler(isArrayUpdate, function (update, change) {
  var propertyPathAsArray = isArrayIndexPath(change.diff.path) ? change.diff.path.slice(0, -1) : change.diff.path;
  var propertyPath = pathToString(propertyPathAsArray);
  setProperty(update, propertyPath, getProperty(change.rhs, propertyPathAsArray));
}), createChangeHandler(function () {
  return true;
}, function (update, change) {
  throw new Error('Unhandled change: ' + (0, _stringify2.default)(change));
})];

var applyTheChangeHandlers = _lodash2.default.partial(applyChangeHandlers, changeHandlers);

function diff(lhs, rhs) {
  var theDiff = (0, _deepDiff.diff)(lhs, rhs);
  return _lodash2.default.reduce(theDiff, function (update, changeDiff) {
    applyTheChangeHandlers(update, { lhs: lhs, rhs: rhs, diff: changeDiff });
    return update;
  }, {});
}
